### PR TITLE
Add DetConS pytorch training example

### PR DESCRIPTION
**Pull request by saud5150**
Part of #1957

## Summary

- Adds `examples/pytorch/detcon.py` for DetConS (SimCLR-based contrastive detection).
- Uses `DetConSTransform(grid_size=(5, 5))` with the built-in `AddGridTransform` — no
  external segmentation dependency (no scikit-image needed).

## Testing

- `make format`
- `make generate-example-notebooks`
- `make all-checks` *(fails locally in `tests/utils/test_dist__gather__losses.py::TestGatherLayer_Losses::test_loss_dcl` and 3 others; all pre-exist on upstream master)*
- Trained the example for 5 epochs on CIFAR-10; loss went from 18.5 to 15.3.

### CODE REVIEW

**Review by gabrielfruet — APPROVED** [review]
LGTM, Thank you for your nice contribution!

**Comment by liopeer** [comment]
Thanks a lot @saud5150! I saw that you use grid masks in your examples. Would it be possible to extend the example unsupervised masks like in the paper? We can also add an optional dependency like scikit-image to make this work. I would imagine something like the following:

```python
# in order to run unsupervised mask segmentation with this example make sure to also install
# scikit-image, i.e. run `pip install lightly scikit-image`

try:
    from skimage.segmentation import felzenszwalb
    SCIKIT_IMAGE_INSTALLED = True
except ImportError:
    print("Scikit-image is not installed, running the example with grid masks.")
    SCIKIT_IMAGE_INSTALLED = False

# ...
if SCIKIT_IMAGE_INSTALLED:
    _detcons_transform = DetConSTransform(input_size=96) # no `grid_size`
else:
    # existing transform

# in the training loop, calculate the mask
```
It would be better if the DetConS transform actually support passing a `callable` to generate the mask, so that it could be done in the dataloader workers, but I would leave that for another day.

**Comment by saud5150** [comment]
> Thanks a lot @saud5150! I saw that you use grid masks in your examples. Would it be possible to extend the example unsupervised masks like in the paper? We can also add an optional dependency like scikit-image to make this work. I would imagine something like the following:
> 
> ```python
> # in order to run unsupervised mask segmentation with this example make sure to also install
> # scikit-image, i.e. run `pip install lightly scikit-image`
> 
> try:
>     from skimage.segmentation import felzenszwalb
>     SCIKIT_IMAGE_INSTALLED = True
> except ImportError:
>     print("Scikit-image is not installed, running the example with grid masks.")
>     SCIKIT_IMAGE_INSTALLED = False
> 
> # ...
> if SCIKIT_IMAGE_INSTALLED:
>     _detcons_transform = DetConSTransform(input_size=96) # no `grid_size`
> else:
>     # existing transform
> 
> # in the training loop, calculate the mask
> ```
> 
> It would be better if the DetConS transform actually support passing a `callable` to generate the mask, so that it could be done in the dataloader workers, but I would leave that for another day.

Implemented try/except with Felzenszwalb as primary path and the grid as fallback. Used scale=100, sigma=0.5, min_size=20 which gives ~10–15 segments on 32×32 CIFAR images; clipped output to num_cls=25 so model architecture stays the same either way. Happy to tune the parameters.

**Comment by liopeer** [comment]
I think that sounds reasonable, thanks a lot! Will you post the implementation together with the PRs for the Lightning examples? @saud5150

**Comment by saud5150** [comment]
Yes, I'll open a follow-up PR for the scikit-image path in the pytorch example (since #1961 is already merged), and the Lightning examples will include it from the start. Will have both up shortly.
